### ISSUE TITLE
Crash in AWSMobileClient.releaseSignInWait()

### ISSUE THREAD

**Issue by alrashedf** [aws-amplify/aws-sdk-ios#3976]
**Describe the bug**
The app require logging in with Cognito. When the app first starts, it crashes with the stack trace below. 

_Observation_: What I don't understand is that there are two `self.tokenFetchLock.leave()` calls in the SDK but no corresponding  `self.tokenFetchLock.enter()`.  `self.tokenFetchLock` seems to be useless and should be removed, right?

**To Reproduce**
Running the app is enough to produce the bug. The Cognito token timeout is set at 5 minutes to make the bug more pronounced.

**Observed Behavior**
Crash

**Expected Behavior**
Normal continuation of app as a logged in user.

**Stack Trace**

```
Hardware Model:      iPhone13,3
AppStoreTools:       13C90b
AppVariant:          1:iPhone13,3:15
Beta:                YES
Code Type:           ARM-64 (Native)
Role:                Foreground
Parent Process:      launchd [1]
Coalition:           com.asusual.ios [3773]

Date/Time:           2022-01-18 17:42:30.0553 +0300
Launch Time:         2022-01-18 17:42:28.4074 +0300
OS Version:          iPhone OS 15.1.1 (19B81)
Baseband Version:    2.11.04
Report Version:      104

Exception Type:  EXC_BREAKPOINT (SIGTRAP)
Exception Codes: 0x0000000000000001, 0x0000000181919a64
Exception Note:  EXC_CORPSE_NOTIFY
Terminating Process: exc handler [11678]
Triggered by Thread:  0


Kernel Triage:
VM - pmap_enter failed with resource shortage
VM - pmap_enter failed with resource shortage
VM - pmap_enter failed with resource shortage
VM - pmap_enter failed with resource shortage
VM - pmap_enter failed with resource shortage


Thread 0 name:
Thread 0 Crashed:
0   libdispatch.dylib             	0x0000000181919a64 dispatch_group_leave.cold.1 + 36 (semaphore.c:303)
1   libdispatch.dylib             	0x00000001818e44c8 dispatch_group_leave + 140 (semaphore.c:0)
2   AWSMobileClient               	0x0000000105396f88 AWSMobileClient.releaseSignInWait() + 1052 (AWSMobileClientExtensions.swift:826)
3   AmplifyPlugins                	0x000000010591d5fc AWSMobileClientAdapter.releaseSignInWait() + 8 (AWSMobileClientAdapter.swift:258)
4   AmplifyPlugins                	0x000000010591d5fc protocol witness for AWSMobileClientBehavior.releaseSignInWait() in conformance AWSMobileClientAdapter + 12 (<compiler-generated>:257)
5   AmplifyPlugins                	0x000000010591d5fc closure #1 in AuthorizationProviderAdapter.setupListener() + 164 (AuthorizationProviderAdapter.swift:59)
6   AWSMobileClient               	0x0000000105389f14 thunk for @escaping @callee_guaranteed (@unowned UserState, @guaranteed [String : String]) -> () + 12 (<compiler-generated>:0)
7   AWSMobileClient               	0x0000000105389f14 partial apply for thunk for @escaping @callee_guaranteed (@unowned UserState, @guaranteed [String : String]) -> () + 28
8   AWSMobileClient               	0x0000000105397518 thunk for @escaping @callee_guaranteed (@in_guaranteed UserState, @in_guaranteed [String : String]) -> (@out ()) + 12 (<compiler-generated>:0)
9   AWSMobileClient               	0x0000000105397518 AWSMobileClient.mobileClientStatusChanged(userState:additionalInfo:) + 12
10  AWSMobileClient               	0x0000000105397518 closure #1 in closure #1 in AWSMobileClient.getTokens(_:) + 704
11  AWSMobileClient               	0x0000000105397518 0x105368000 + 193816 (AWSMobileClient.swift:377)
12  AWSMobileClient               	0x0000000105397518 0x105368000 + 193816 (AWSMobileClientExtensions.swift:857)
13  AWSMobileClient               	0x0000000105385ed0 thunk for @escaping @callee_guaranteed (@guaranteed AWSCognitoAuthUserSession?, @guaranteed Error?) -> () + 68 (<compiler-generated>:0)
14  AWSMobileClient               	0x0000000105373b14 -[AWSCognitoAuth cleanUpAndCallGetSessionBlock:error:] + 116 (AWSCognitoAuth.m:508)
15  AWSMobileClient               	0x0000000105373994 -[AWSCognitoAuth dismissSafariViewControllerAndCompleteGetSession:error:] + 140 (AWSCognitoAuth.m:496)
16  AWSMobileClient               	0x00000001053767bc -[AWSCognitoAuth connectionDidFinishLoading:] + 244 (AWSCognitoAuth.m:0)
17  CFNetwork                     	0x0000000182689920 __65-[NSURLConnectionInternal _withConnectionAndDelegate:onlyActive:]_block_invoke + 60 (NSURLConnectionInternal.mm:122)
18  CFNetwork                     	0x00000001826898c0 -[NSURLConnectionInternal _withConnectionAndDelegate:onlyActive:] + 188 (NSURLConnectionInternal.mm:119)
19  CFNetwork                     	0x00000001826899e0 -[NSURLConnectionInternal _withActiveConnectionAndDelegate:] + 56 (NSURLConnectionInternal.mm:137)
20  CFNetwork                     	0x00000001825c8994 invocation function for block in URLConnectionClient_Classic::_delegate_didFinishLoading(void () block_pointer) + 52 (URLConnectionClient.cpp:1873)
21  CFNetwork                     	0x00000001825c71e8 invocation function for block in URLConnectionClient_Classic::_withDelegateAsync(char const*, void (_CFURLConnection*, CFURLConnectionClientCurrent_VMax const*) block_pointer) + 136 (URLConnectionClient.cpp:0)
22  libdispatch.dylib             	0x00000001818e3660 _dispatch_client_callout + 20 (object.m:560)
23  libdispatch.dylib             	0x00000001818e7118 _dispatch_block_invoke_direct + 264 (queue.c:489)
24  CFNetwork                     	0x000000018266a650 RunloopBlockContext::_invoke_block(void const*, void*) + 44 (CoreSchedulingSet.mm:365)
25  CoreFoundation                	0x0000000181be53fc CFArrayApplyFunction + 84 (CFArray.c:699)
26  CFNetwork                     	0x000000018266a540 RunloopBlockContext::perform() + 136 (CoreSchedulingSet.mm:319)
27  CFNetwork                     	0x000000018266caac MultiplexerSource::_perform(void*) + 336 (CFNRunLoopMultiplexer.c:267)
28  CoreFoundation                	0x0000000181c93020 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 28 (CFRunLoop.c:1972)
29  CoreFoundation                	0x0000000181ca3ce0 __CFRunLoopDoSource0 + 208 (CFRunLoop.c:2016)
30  CoreFoundation                	0x0000000181bddfe8 __CFRunLoopDoSources0 + 268 (CFRunLoop.c:2053)
31  CoreFoundation                	0x0000000181be37f4 __CFRunLoopRun + 820 (CFRunLoop.c:2951)
32  CoreFoundation                	0x0000000181bf73b8 CFRunLoopRunSpecific + 600 (CFRunLoop.c:3268)
33  GraphicsServices              	0x000000019d58738c GSEventRunModal + 164 (GSEvent.c:2200)
34  UIKitCore                     	0x00000001845976a8 -[UIApplication _run] + 1100 (UIApplication.m:3493)
35  UIKitCore                     	0x00000001843167f4 UIApplicationMain + 2092 (UIApplication.m:5046)
36  Runner                        	0x00000001042affa8 main + 64 (AppDelegate.swift:28)
37  dyld                          	0x0000000104c2da24 start + 520 (dyldMain.cpp:876)
```

**Code Snippet**
```
    public func releaseSignInWait() {
        if self.federationProvider == .userPools {
            self.userpoolOpsHelper.passwordAuthTaskCompletionSource?.set(error: AWSMobileClientError.unableToSignIn(message: "Unable to get valid sign in session from the end user."))
            self.userpoolOpsHelper.passwordAuthTaskCompletionSource = nil
            self.userpoolOpsHelper.customAuthChallengeTaskCompletionSource?.set(error: AWSMobileClientError.unableToSignIn(message: "Unable to get valid sign in session from the end user."))
            self.userpoolOpsHelper.customAuthChallengeTaskCompletionSource = nil
        } else if self.federationProvider == .hostedUI {
            self.pendingGetTokensCompletion?(nil, AWSMobileClientError.unableToSignIn(message: "Could not get valid token from the user."))
            self.pendingGetTokensCompletion = nil
            self.tokenFetchLock.leave() <------------------- crash here
        } else if self.federationProvider == .oidcFederation {
            self.pendingAWSCredentialsCompletion?(nil, AWSMobileClientError.unableToSignIn(message: "Could not get valid federation token from the user."))
            self.pendingAWSCredentialsCompletion = nil
        }
    }
```

**Areas of the SDK you are using (AWSMobileClient, Cognito, Pinpoint, IoT, etc)?**
Cognito, DataStore APIs

**Environment(please complete the following information):**
 - SDK Version: 2.26.6
 - Dependency Manager: Cocoapods
 - Swift Version : 5.5.2-dev
 - Xcode Version: 13.2.1 (13C100)

**Device Information (please complete the following information):**
 - Device: all devices
 - iOS Version: 15.1.1

**Additional context**
I'm using:
[✓] Flutter (Channel stable, 2.8.1, on macOS 12.1 21C52 darwin-x64, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
[✓] Android Studio (version 2020.3)

**Comment by alrashedf** [comment]
It seems I missed that a fix is already done #3951

**Comment by harsh62** [comment]
Since the issue already has a fix. Closing the issue.
